### PR TITLE
[clang][CodeGen] Emit atomic IR in place of optimized libcalls.

### PR DESCRIPTION
**Pull request by Logikable**
In the beginning, Clang only emitted atomic IR for operations it knew the
underlying microarch had instructions for, meaning it required significant
knowledge of the target. Later, the backend acquired the ability to lower
IR to libcalls. To avoid duplicating logic and improve logic locality,
we'd like to move as much as possible to the backend.

There are many ways to describe this change. For example, this change
reduces the variables Clang uses to decide whether to emit libcalls or
IR, down to only the atomic's size.

### CODE REVIEW

**Comment by arichardson** [comment]
When I first came across this code I found it quite surprising that both clang and LLVM need handling for atomic libcalls. I feel it would be a lot nicer if this could all be handled inside the LLVM backend logic. Would it also be possible to emit the unoptimized libcall from LLVM instead of Clang and always emit the atomic IR operation and let the backend deal with lowering or am I missing something?

**Comment by Logikable** [comment]
I believe Clang currently does not properly propagate alignment information for certain IR (e.g. atomicrmw and cmpxchg). Look at the definitions of `CreateAtomic*` in Clang. The patches in https://bugs.llvm.org/show_bug.cgi?id=27168 did not address this (thanks @MaskRay for helping me find this).

**Comment by efriedma-quic** [comment]
As a historical note, when I first wrote the support for atomics, it was under a model where LLVM IR only modeled legal atomics, and clang lowered illegal atomics to libcalls.  This was changed a few years later to generalize atomics in LLVM IR to support arbitrary atomic ops lowered to libcalls.  But there are still a few inconsistencies resulting from this, I think.

**Comment by Logikable** [comment]
I've rewritten a part of this PR to incorporate @arichardson's suggestion. I wasn't able to find inconsistencies through a cursory review of the backend's AtomicExpand pass, nor through a small set of tests built for common targets.

How can I find the gaps I missed?

**Comment by efriedma-quic** [comment]
When I said "inconsistencies", I just meant the way the responsibility for lowering atomics is split between LLVM and clang; I didn't mean anything was actually broken.

**Review by jyknight — COMMENTED** [review]
Overall, I think this is a great improvement.

And I believe it should be correct to do now, after fixing the backends' MaxAtomicSizeInBitsSupported in #75703 #75185 #75112 #74385 #74389, and fixing Clang's propagation of alignment to atomicrmw/cmpxchg in #74349.

Just a few more comments.

File: clang/test/CodeGen/LoongArch/atomics.c [hunk 1]
```diff
@@ -33,9 +33,9 @@ void test_i32_atomics(_Atomic(int32_t) * a, int32_t b) {
...
 void test_i64_atomics(_Atomic(int64_t) * a, int64_t b) {
-  // LA32: call i64 @__atomic_load_8
-  // LA32: call void @__atomic_store_8
-  // LA32: call i64 @__atomic_fetch_add_8
+  // LA32: load atomic i64, ptr %a seq_cst, align 8
+  // LA32: store atomic i64 %b, ptr %a seq_cst, align 8
+  // LA32: atomicrmw add ptr %a, i64 %b seq_cst
   // LA64: load atomic i64, ptr %a seq_cst, align 8
   // LA64: store atomic i64 %b, ptr %a seq_cst, align 8
   // LA64: atomicrmw add ptr %a, i64 %b seq_cst
```
Comment (arichardson) [inline clang/test/CodeGen/LoongArch/atomics.c:41]: Also missing alignment here

File: clang/lib/CodeGen/CGAtomic.cpp [hunk 2]
```diff
@@ -1047,122 +1019,19 @@ RValue CodeGenFunction::EmitAtomicExpr(AtomicExpr *E) {
...
-    case AtomicExpr::AO__scoped_atomic_exchange_n:
-    case AtomicExpr::AO__scoped_atomic_compare_exchange_n:
-      // Only use optimized library calls for sizes for which they exist.
-      // FIXME: Size == 16 optimized library functions exist too.
-      if (Size == 1 || Size == 2 || Size == 4 || Size == 8)
-        UseOptimizedLibcall = true;
-      break;
-    }
+  bool PowerOf2Size = (Size & (Size - 1)) == 0;
+  bool UseLibcall = !PowerOf2Size | (Size > 16);
```
[the remark this answers is not in the document]
Comment (Logikable) [reply clang/lib/CodeGen/CGAtomic.cpp:1023]: I've changed `|` -> `||`.

`&` is part of the operation here (I missed that initially too when reading the comment).

**Review by arichardson — COMMENTED** [review]
Overall this looks great to me, very happy to see reduced duplication between clang and llvm.

It would be great if we could just emit the atomic IR in all cases but unfortunately non-power-of-two sizes result in a verifier error. We could probably emit the atomic IR for power-of-two sizes and let the expand pass deal with it but since we still need the fallback for other sizes we might as well keep it.

**Comment by Logikable** [comment]
Updated with suggestions. How does this look?

**Comment by jyknight** [comment]
BTW, please don't rebase/amend commits and force-push, it makes a it extremely difficult to review the changes since the previous review -- especially if the new changes are commingled with a rebase.

Much better to simply push new commits on top of your existing branch. And if you need an updated baseline, you can `git merge origin/main` in your PR branch.

**Review by jyknight — APPROVED** [review]
Sorry for the delay. This looks good to me now!

**Comment by rorth** [comment]
I'm pretty certain this patch broke the [Solaris/sparcv9 buildbot](https://lab.llvm.org/buildbot/#/builders/72/builds/2541).

**Comment by rorth** [comment]
Confirmed: reverting the patch locally restores the build.

**Comment by Logikable** [comment]
https://reviews.llvm.org/D118021 likely related, looking into this.

**Comment by rorth** [comment]
The libcalls per se wouldn't be a problem since on 32-bit Solaris/sparc `-latomic` is always linked due to that.  However, the `__sync_*` libcalls aren't defined in `libatomic.so`, thus the link failure.

**Comment by efriedma-quic** [comment]
Looks like a bug in the SPARC backend.  https://github.com/llvm/llvm-project/pull/73176#pullrequestreview-1811677691  indicated that atomic expansion was working correctly for all backends, but I guess it isn't working correctly on SPARC.

**Comment by jyknight** [comment]
What config is this bot using? Is it "-target sparc-solaris -mcpu=v9" (that is: 32-bit sparc, but with v9 cpu available)?

I see that SparcV9TargetInfo sets MaxAtomicInlineWidth to 64 unconditionally. But, the message `warning: large atomic operation may incur significant performance penalty; the access size (8 bytes) exceeds the max lock-free size (4  bytes) [-Watomic-alignment]` indicates that MaxAtomicInlineWidth is 32. OK, wait, this class is poorly-named. SparcV8TargetInfo really means "sparc 32-bit" and SparcV9TargetInfo really means "sparc 64-bit". So: this is actually using SparcV8TargetInfo, instead. 

Yet, even then, SparcV8TargetInfo appears to [intend to configure](https://github.com/llvm/llvm-project/blob/16140ff219b68f61fedf92df13019d89a4990a47/clang/lib/Basic/Targets/Sparc.h#L168) MaxAtomicInlineWidth to 64 when using a V9 CPU. That doesn't seem to be actually happening, though.

OK, so the frontend is apparently buggily configured to only support 32-bit atomics in 32-bit mode, even when a V9 CPU is available.

That, then, was covering up a bug in the backend. The backend claims to (and should be able to!) support 64-bit atomics in 32-bit mode. But it doesn't actually, right now.

I'll send a patch to fix this by limiting atomics to 32-bit in the backend when `!Subtarget->is64Bit()`, which will get things back to working, though it doesn't address those deeper issues.

**Comment by jrtc27** [comment]
Part of that confusion comes from SPARC's own naming. V9 is the CPU, but a V9 CPU being used for 32-bit code is called V8+...

**Pull request merged**